accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -699,11 +702,27 @@ the element, key, and value type of `...e` and `...?e` is `Never`.
 *When the static type _S_ of `e` is strictly non-nullable, such as when _S_
 is `Never`, `...?e` is a warning, but it may still occur.*
 
-### Instantiate to bounds
+### Instantiation to bound
 
-The computation of instantiation to bounds is changed to substitute `Never` for
+The computation of instantiation to bound is changed to substitute `Never` for
 type variables appearing in contravariant positions instead of `Null`.
 
+### Super-bounded types
+
+Null safety requires two changes to the section 'Super-Bounded Types' in the
+language specification.
+
+The definition of a top type is changed: _T_ is a top type if and only if
+`Object?` is a subtype of _T_.
```

Comment:
Is this the same as the `TOP` predicate declared [here](https://github.com/dart-lang/language/blob/master/resources/type-system/upper-lower-bounds.md#helper-predicates)?

Reply:
This document already uses 'top type' several times, and does not refer to **TOP**, so we should in any case resolve that question.

I think **TOP** can be simplified to use the definition of 'top type' given here. If we have a top type `T`, then **TOP**(`T`) holds:

- If we used Reflexivity then we know that `T == Object?`
    OK: `TOP(Object?)` because `OBJECT(Object)`.
- Right top: `T == dynamic/void/Object?`
    OK: `TOP(dynamic)`, `TOP(void)`, `TOP(Object?)`.
- Right legacy, when `T == S*`: `Object? <: S?`
    OK: `TOP(S*)` by lemma 2, 1.
- Left nullable: `Object <: T && Null <: T`
    OK: `TOP(T)` by lemma 2: If `TOP(T)` we're done, if `OBJECT(T)` then
       `T == FutureOr^k<Object>`, contradicting `Null <: T`.
- Right FutureOr, when `T == FutureOr<S>`: `Object? <: S`
    OK: by induction `TOP(S)`, so `TOP(FutureOr<S>)`.
- Right Nullable, when `T == S?`: `Object? <: S`
    OK: by induction `TOP(S)`, so `TOP(S?)`.

Lemma 1: `Object? <: S?` => `Object <: S`
Lemma 2: `Object <: T` => `OBJECT(T) || TOP(T)`

It's rather easy to see that if `TOP(T)` then `T` is a top type.

So we should actually be able simplify `TOP`.

Reply:
Pull helper predicates to the null safety feature spec (TOP, BOTTOM, ...) and refer to them from upper-lower-bounds.md